File: scripts/extract-text-coords.py
"""
Extract text items with their PDF coordinate positions from briefing-cert-en.pdf.
Uses pdfminer.six which properly decompresses and parses PDF content streams.
PDF coordinate space: origin at bottom-left, y increases upward.
"""
import sys
from pdfminer.high_level import extract_pages
from pdfminer.layout import LTTextBox, LTTextLine, LTChar
from pdfminer.layout import LAParams

pdf_path = "docs/templates/briefing-cert-en.pdf"

print(f"Extracting coordinates from: {pdf_path}\n")

laparams = LAParams(line_margin=0.5, word_margin=0.1)

for page_num, page_layout in enumerate(extract_pages(pdf_path, laparams=laparams), 1):
    page_height = page_layout.height
    page_width = page_layout.width
    print(f"Page {page_num}: {page_width:.1f}pt x {page_height:.1f}pt")
    print(f"(y=0 at bottom; y={page_height:.0f} at top)\n")
    
    # Collect all text lines with positions
    lines = []
    for element in page_layout:
        if isinstance(element, LTTextBox):
            for line in element:
                if isinstance(line, LTTextLine):
                    text = line.get_text().strip()
                    if text:
                        x0, y0, x1, y1 = line.bbox
                        lines.append((y0, x0, x1, y1, text))
    
    # Sort by y descending (top of page first)
    lines.sort(key=lambda t: -t[0])
    
    print("=== ALL TEXT LINES (y from bottom, top of page first) ===")
    for y0, x0, x1, y1, text in lines:
        print(f"  y={y0:6.1f} | x={x0:6.1f}..{x1:5.1f} | {text}")
    
    print("\n=== KEYWORD MATCHES (name/signature/date/household) ===")
    keywords = ['name', 'signature', 'date', 'head', 'household', 'printed', 'sign']
    for y0, x0, x1, y1, text in lines:
        if any(kw in text.lower() for kw in keywords):
            print(f"  *** y={y0:6.1f} | x={x0:6.1f} | {text}")

File: scripts/extract-text-coords.mjs
/**
 * Extract text items with their PDF coordinate positions from briefing-cert-en.pdf.
 * Uses pdfjs-dist (already installed) which decompresses and parses the content streams.
 * The transform matrix [a,b,c,d,e,f] where e=x, f=y in PDF user space.
 * pdfjs uses top-left origin; we convert to PDF bottom-left for pdf-lib compatibility.
 */
import { createRequire } from 'module';
import { readFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const root = join(__dirname, '..');
const require = createRequire(import.meta.url);

// Use legacy CJS build
const pdfjs = require('pdfjs-dist/legacy/build/pdf.js');

const pdfPath = join(root, 'docs/templates/briefing-cert-en.pdf');
const data = new Uint8Array(readFileSync(pdfPath));

const loadingTask = pdfjs.getDocument({ data });
const pdf = await loadingTask.promise;

console.log(`PDF loaded. Pages: ${pdf.numPages}`);

const page = await pdf.getPage(1);
const viewport = page.getViewport({ scale: 1.0 });
const pageHeight = viewport.height; // in PDF points

console.log(`Viewport: ${viewport.width}pt x ${viewport.height}pt\n`);

const textContent = await page.getTextContent();
const items = textContent.items;

console.log('=== ALL TEXT ITEMS WITH COORDINATES ===');
console.log('Format: y_from_bottom | x | text');
console.log('(y=0 at bottom; typical letter page = 790pt tall)\n');

// Group by approximate Y (within 3pt = same line)
const lines = new Map();
for (const item of items) {
  if (!item.str || item.str.trim() === '') continue;
  // pdfjs transform[4]=x, transform[5]=y (from bottom in PDF space)
  const x = item.transform[4];
  const y = item.transform[5];
  const yKey = Math.round(y / 3) * 3; // bucket to 3pt groups
  if (!lines.has(yKey)) lines.set(yKey, []);
  lines.get(yKey).push({ x, y, text: item.str });
}

// Sort by y descending (top of page first)
const sorted = [...lines.entries()].sort((a, b) => b[0] - a[0]);

for (const [yKey, lineItems] of sorted) {
  // Sort items in line by x
  lineItems.sort((a, b) => a.x - b.x);
  const lineText = lineItems.map(i => i.str).join('');
  const y = lineItems[0].y;
  const x = lineItems[0].x;
  console.log(`y=${y.toFixed(1).padStart(6)} | x=${x.toFixed(1).padStart(6)} | ${lineText}`);
}

console.log('\n=== SIGNATURE / DATE AREA — look for lines near bottom of form ===');
console.log('Target lines containing "Name", "Signature", "Date", "Head", "Household"');
const keywords = ['name', 'signature', 'date', 'head', 'household', 'printed', 'sign'];
for (const [yKey, lineItems] of sorted) {
  lineItems.sort((a, b) => a.x - b.x);
  const lineText = lineItems.map(i => i.str).join('').toLowerCase();
  if (keywords.some(kw => lineText.includes(kw))) {
    const y = lineItems[0].y;
    const x = lineItems[0].x;
    const display = lineItems.map(i => i.str).join('');
    console.log(`  y=${y.toFixed(1).padStart(6)} | x=${x.toFixed(1).padStart(6)} | >>> ${display} <<<`);
  }
}
